File: src/parsing/functions/min.ts
import AggregateFunction from "./aggregate_function";
import { FunctionDef } from "./function_metadata";
import ReducerElement from "./reducer_element";

class MinReducerElement extends ReducerElement {
    private _value: any = null;
    public get value(): any {
        return this._value;
    }
    public set value(value: any) {
        if (this._value === null || value < this._value) {
            this._value = value;
        }
    }
}

@FunctionDef({
    description: "Returns the minimum value across grouped rows",
    category: "aggregate",
    parameters: [{ name: "value", description: "Value to compare", type: "number" }],
    output: { description: "Minimum value", type: "number", example: 1 },
    examples: ["WITH [3, 1, 2] AS nums UNWIND nums AS n RETURN min(n)"],
})
class Min extends AggregateFunction {
    constructor() {
        super("min");
        this._expectedParameterCount = 1;
    }
    public reduce(element: MinReducerElement): void {
        element.value = this.firstChild().value();
    }
    public element(): MinReducerElement {
        return new MinReducerElement();
    }
}

export default Min;

File: flowquery-py/src/parsing/functions/min_.py
"""Min aggregate function."""

from typing import Any

from .aggregate_function import AggregateFunction
from .function_metadata import FunctionDef
from .reducer_element import ReducerElement


class MinReducerElement(ReducerElement):
    """Reducer element for Min aggregate function."""

    def __init__(self) -> None:
        self._value: Any = None

    @property
    def value(self) -> Any:
        return self._value

    @value.setter
    def value(self, val: Any) -> None:
        if self._value is None or val < self._value:
            self._value = val


@FunctionDef({
    "description": "Returns the minimum value across grouped rows",
    "category": "aggregate",
    "parameters": [
        {"name": "value", "description": "Value to compare", "type": "number"}
    ],
    "output": {"description": "Minimum value", "type": "number", "example": 1},
    "examples": ["WITH [3, 1, 2] AS nums UNWIND nums AS n RETURN min(n)"]
})
class Min(AggregateFunction):
    """Min aggregate function.

    Returns the minimum value across grouped rows.
    """

    def __init__(self) -> None:
        super().__init__("min")
        self._expected_parameter_count = 1

    def reduce(self, element: MinReducerElement) -> None:
        element.value = self.first_child().value()

    def element(self) -> MinReducerElement:
        return MinReducerElement()
